RandLAPACK/gpu_functions/rl_cuda_kernels.cuh

```diff
@@ -0,0 +1,934 @@
+        abort();
+    }
+}
+
+/// Positions columns of A in accordance with idx vector of length k.
+/// idx array modified ONLY within the scope of this function.
+template <typename T>
+void col_swap_gpu(
+    cudaStream_t stream, 
+    int64_t m,  
+    int64_t k,
+    int64_t* J, 
+    int64_t const* idx
+) {
+#if 1
+    std::vector<int64_t> idx_copy(k);
+    auto j = std::make_unique<int64_t[]>(m);
+    cudaMemcpyAsync(idx_copy.data(), idx, sizeof(int64_t) * k, cudaMemcpyDeviceToHost, stream);
+    cudaMemcpyAsync(j.get(), J, sizeof(int64_t) * m, cudaMemcpyDeviceToHost, stream);
+    cudaStreamSynchronize(stream);
+
+    RandLAPACK::util::col_swap<T>(m, k, j.get(), std::move(idx_copy));
+
+    cudaMemcpyAsync(J, j.get(), sizeof(int64_t) * m, cudaMemcpyHostToDevice, stream);
+    // must add this to avoid dangling reference during async copy
+    cudaStreamSynchronize(stream);
+#else
+#ifdef USE_CUDA
+    int64_t* idx_copy;
+    cudaMallocAsync(&idx_copy, sizeof(int64_t) * k, stream);
+    vec_ell_swap_gpu<T><<<1, 512, 0, stream>>>(m, k, J, idx, idx_copy);
+#endif
+    cudaFreeAsync(idx_copy, stream);
+    cudaError_t ierr = cudaGetLastError();
+    if (ierr != cudaSuccess)
+    {
+        RandLAPACK_CUDA_ERROR("Failed to launch col_swap_gpu. " << cudaGetErrorString(ierr))
+        abort();
+    }
+#endif
+}
+
+template <typename T>
+void col_swap_gpu(
+    cudaStream_t stream, 
+    int64_t m,  
+    int64_t k,
+    int64_t* J, 
+    int64_t* J_cpy, 
+    int64_t const* idx
+) {
+#ifdef USE_CUDA
+    constexpr int threadsPerBlock{128};
+    int64_t num_blocks{(k + threadsPerBlock - 1) / threadsPerBlock};
+    vec_ell_swap_gpu<T><<<num_blocks, threadsPerBlock, 0, stream>>>(m, k, J, J_cpy, idx);
+#endif
+    cudaError_t ierr = cudaGetLastError();
+    if (ierr != cudaSuccess)
+    {
+        RandLAPACK_CUDA_ERROR("Failed to launch col_swap_gpu vector. " << cudaGetErrorString(ierr))
+        abort();
+    }
+}
+
+
+template <typename T>
+void transposition_gpu(
+    cudaStream_t stream, 
+    int64_t m,
+    int64_t n,
+    const T* A,
+    int64_t lda,
+    T* AT,
+    int64_t ldat,
+    bool copy_upper_triangle
+) {
+#ifdef USE_CUDA
+    constexpr int threadsPerBlock{128};
+    int64_t num_blocks{(m + threadsPerBlock - 1) / threadsPerBlock};
+    transposition_gpu<<<num_blocks, threadsPerBlock, 0, stream>>>(m, n, A, lda, AT, ldat, copy_upper_triangle);
+#endif
+    cudaError_t ierr = cudaGetLastError();
+    if (ierr != cudaSuccess)
+    {
+        RandLAPACK_CUDA_ERROR("Failed to launch transposition_gpu. " << cudaGetErrorString(ierr))
+        abort();
+    }
+}
+
+
+// This function is the GPU equivalent of the operations performed on the pivot vector
+// in the context of PLU-based QRCP.
+// The implementation is very clumsy, but we needed to put it on a GPU.
+// TODO: we get a multiple definition linker error if this function is not templated.
+template <typename Idx>
+inline void LUQRCP_piv_porcess_gpu(
```

Comment:
Please do a CTRL+R, replacing "porcess" with "process" throughout this file and all others.

Reply:
Fixed.